src/ddprof_process.cc

```diff
@@ -0,0 +1,143 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0. This product includes software
+// developed at Datadog (https://www.datadoghq.com/). Copyright 2021-Present
+// Datadog, Inc.
+
+#include "ddprof_process.hpp"
+
+#include "ddres.hpp"
+#include "string_format.hpp"
+
+#include <charconv> // for std::from_chars
+#include <unistd.h>
+#include <vector>
+
+#include <fstream>
+#include <optional>
+#include <regex>
+#include <string>
+
+namespace ddprof {
+constexpr auto k_max_buf_cgroup_link = 1024;
+
+constexpr static std::string_view UUID_SOURCE =
+    "[0-9a-f]{8}[-_][0-9a-f]{4}[-_][0-9a-f]{4}[-_][0-9a-f]{4}[-_][0-9a-f]{12}";
+constexpr static std::string_view CONTAINER_SOURCE = "[0-9a-f]{64}";
+constexpr static std::string_view TASK_SOURCE = "[0-9a-f]{32}-\\d+";
+
+namespace {
+std::optional<std::string> container_id_from_line(const std::string &line) {
+  std::smatch line_matches;
+  std::smatch container_matches;
+  const std::regex LINE_REGEX(R"(^\d+:[^:]*:(.+)$)");
+  const std::regex CONTAINER_REGEX(
+      "(" + std::string(UUID_SOURCE) + "|" + std::string(CONTAINER_SOURCE) +
+      "|" + std::string(TASK_SOURCE) + ")(?:.scope)? *$");
+
+  if (std::regex_search(line, line_matches, LINE_REGEX)) {
+    std::string match = line_matches[1].str();
+    if (std::regex_search(match, container_matches, CONTAINER_REGEX)) {
+      return container_matches[1].str();
+    }
+  }
+  return std::nullopt;
+}
+
+} // namespace
+
+std::string Process::format_cgroup_file(pid_t pid,
+                                        std::string_view path_to_proc) {
+  return string_format("%s/proc/%d/cgroup", path_to_proc.data(), pid);
+}
+
+DDRes extract_container_id(const std::string &filepath,
+                           ContainerId &container_id) {
+  container_id = std::nullopt;
+  std::ifstream cgroup_file(filepath);
+  if (!cgroup_file) {
+    // short lived pids can fail in this case
+    LG_DBG("Failed to open file: %s", filepath.data());
+    return ddres_warn(DD_WHAT_CGROUP);
+  }
+  std::string line;
+  while (std::getline(cgroup_file, line)) {
+    container_id = container_id_from_line(line);
+    if (container_id) {
+      return {};
+    }
+  }
+  container_id = "none";
```

Comment:
Ah, interesting.  So our default is `"unknown"`, but failure to process this upgrades it to `"none"`.  I think there's value in having a semantic distinction and I think these two values are as good as any other, but in the future we may want to be more descriptive.

Reply:
Thanks for highlighting that this was unclear. I added comments.
`none` means that it is operating outside a container.